lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -1997,6 +1988,7 @@ impl<'a> FuncGen<'a> {
 
         if self.unreachable_depth > 0 {
             was_unreachable = true;
+            self.consume_gas_offset(); // do not instrument unreachable code
```

Comment:
Nice catch! This triggered tests to show me that the `end` in `(loop (br 0) end)` was considered by both unreachability computations as being reachable, but I actually ended up consuming the gas offset too early and so not charging gas for it.

In practice gas still won’t be charged for it because (its cost is 0 and) the label is actually output after the gas instrumentation, but in practice AFAICT the behavior is still fine. Thinking of labels making us jump above instrumentation points led me to check other situations, but I think we’re fine: eg. in a loop we charge for the loop cost on the first loop’s instruction, so even if the label does skip the before-loop charge it actually should, and it’ll get charged the after-loop cost.

However, I don’t think we can actually deploy the asserts to production: it turns out there remain quite a few tests failing due to `(func (unreachable))` having a trailing `end` that’s considered as `reachable` by wasmer and `unreachable` by finite-wasm. Finite-wasm is right here (there’s no branch that’d allow executing it), but I don’t think it’d make sense to make wasmer’s reachability accounting more complex in order to work that out.

(Actually, I’d even have liked to remove all of wasmer’s reachability accounting and force it to always generate the opcodes that are in the wasm (wasm-opt should strip all unreachable code already anyway), even though that’s impossible due to the fact unreachable is stack-polymorphic.)

That said I’m leaving the code as comments in case you have further ideas for how to work that out.